src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,363 @@
+                }
+                else
+                {
+                    // Will be executed in case timer ran out or node is being shut down.
+                    saveBatch = true;
+                }
+
+                if (saveBatch)
+                {
+                    if (batch.Count != 0)
+                    {
+                        await this.SaveBatchAsync(batch).ConfigureAwait(false);
+
+                        batch.Clear();
+                        this.currentBatchSizeBytes = 0;
+                    }
+
+                    timerTask = null;
+                }
+                else
+                {
+                    // Start timer if it is not started already.
+                    timerTask = timerTask ?? Task.Delay(BatchMaxSaveIntervalSeconds * 1000, this.nodeLifetime.ApplicationStopping);
+                }
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>Checks if repository contains reorged blocks and deletes them; saves batch on top. </summary>
+        /// <param name="batch">List of batched blocks. Cannot be empty.</param>
+        private async Task SaveBatchAsync(List<BlockPair> batch)
+        {
+            this.logger.LogTrace("({0}.{1}:{2})", nameof(batch), nameof(batch.Count), batch.Count);
+
+            List<BlockPair> batchCleared = this.GetBatchWithoutReorgedBlocks(batch);
+
+            ChainedHeader expectedStoreTip = batchCleared.First().ChainedHeader.Previous;
+
+            // Check if block repository contains reorged blocks. If it does - delete them.
+            if (expectedStoreTip.HashBlock != this.StoreTip.HashBlock)
+                await this.RemoveReorgedBlocksFromStoreAsync(expectedStoreTip).ConfigureAwait(false);
+
+            // Save batch.
+            ChainedHeader newTip = batchCleared.Last().ChainedHeader;
+
+            this.logger.LogDebug("Saving batch of {0} blocks, total size: {1} bytes.", batchCleared.Count, this.currentBatchSizeBytes);
+
+            await this.blockRepository.PutAsync(newTip.HashBlock, batchCleared.Select(b => b.Block).ToList()).ConfigureAwait(false);
+
+            this.SetStoreTip(newTip);
+
+            this.logger.LogDebug("Store tip set to '{0}'", this.StoreTip);
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Cleans the batch in a way that all headers from the latest one are consecutive.
+        /// Those that violate consecutiveness are removed.
+        /// </summary>
+        /// <param name="batch">Uncleaned batch that might contain non-consecutive blocks. Cannot be empty.</param>
+        /// <returns>List of consecutive blocks.</returns>
+        private List<BlockPair> GetBatchWithoutReorgedBlocks(List<BlockPair> batch)
+        {
+            this.logger.LogTrace("({0}.{1}:{2})", nameof(batch), nameof(batch.Count), batch.Count);
+
+            // Initialize current with highest block from the batch.
+            BlockPair current = batch.Last();
```

Comment:
why `current`? should this naming be a bit longer?  `highestBlockFromLastBatch` and drop the comment?

Reply:
perhaps not that exact name as I can see it is assigned to later - but something in the current naming is losing the flow and making it tricky to follow what it is doing.

Reply:
>highestBlockFromLastBatch and drop the comment?

since the `current` is only initialized with that value and later it is reassigned changing the name to that won't be correct

Reply:
Yep agree